labs/b5.md

```diff
@@ -0,0 +1,364 @@
+192.168.162.0/24 dev eth1  proto kernel  scope link  src 192.168.162.162
+```
+A packet destined for `8.8.8.8` would be forwarded out eth0, the default gateway.
+A packet destined for `10.0.2.1` would be forwarded according to second entry, out of eth0.
+A packet destined for `10.0.2.254` would be forwarded according to third entry, out of eth0.
+A packet destined for `192.168.162.254` would be forwarded according to the fourth entry, out of eth1.
+
+## Private IP addresses and NAT
+Often you'll see a device with an IP address of `192.168.x.x`, and let's say this device is connected to machine A. Pinging that address from A succeeds.  Your friend on machine B, who communicates with you over the internet but isn't connected to your local network, tries to ping that address as well and fails. Now you might be asking ...
+
+
+![alt text][ytho]
+
+[ytho]: http://i0.kym-cdn.com/entries/icons/original/000/022/978/yNlQWRM.jpg
+
+[private]: https://en.wikipedia.org/wiki/IP_address#Private_addresses
+This is because `192.168.0.0/16` is a private network block. [Certain blocks][private] of IP addresses have been designated by the Internet Engineering Task Force (IETF) to be reserved as private address spaces. Private address are especially useful for devices on a local network that only need to communicate on the local network and do not need a public IP address. This is not to say that if you have a private IP address that you cannot reach the wider net or vice versa.
+
+Network address translation or (NAT) is a procedure by which an IP address is mapped into another network address during routing. For example we could map a private IP address to a public one in order to allow a device on the local private network to communicate with cat picture servers out in the wilds of the web.
+
+Take a look at this diagram
+![alt text][NAT]
+
+[NAT]: http://www.h3c.com.hk/res/201212/03/20121203_1477660_image002_765254_1285_0.png "NAT addressing"
+
+Here the private network is on the left side of the router and the right side is the Internet. The router has an interface with a private IP address `192.168.1.1` and an interface with a public IP address `20.1.1.1`. 
+
+For IP packet 1, a host on the private network with address `192.168.1.3` tries to communicate with server B at public address `1.1.1.2` and creates a packet with the proper destination (dst) and source (src).  The packet arrives at the router and the router performs Source Network Address Translation (SNAT) changing the source IP on the packet to be `20.1.1.1` so that if Source B wants to respond it would send a packet to the router's public address since it cannot reach the host at its private IP address. This can be seen in IP packet 2. The packet arrive at the router again and the router now performs Destination Network Address Translation (DNAT) to change the source ip from the router's public address to the private IP address of the host that sent the packet.
```

Comment:
this might seem like magic to the students if we don't mention that NAT works by shuffling L4 ports around between the router's public and private interface so multiple people can appear to be sharing the same L3 address

Reply:
There is a mention of port shuffling as the mechanism behind NAT and a link that should outline that fact

Reply:
NIT: `The packet arrive at the router again` -> `The packet arrives at the router again`